### DISCUSSION TITLE
Transform unavailable when model was fit with only a single data sample.

### DISCUSSION THREAD

**Discussion by mzhadigerov** [MaartenGr/BERTopic#2053] [Q&A] [answered]
Hi! I have 2 lists with strings:

```python
main = [...]
secondary = [...]
```

I want to use the main list to extract topics, then use those topics in a `zeroshot_topic_list` while training a model on `secondary` list.

I could train a model on the `main` list successfully, then I extracted generated topics from the trained model.

But when I pass those topics to train a new model with zero-shot, I'm getting the error:

```
ValueError: Transform unavailable when model was fit with only a single data sample.
```

There are around 53 strings in `main` and around 87 strings in `secondary`.

Here is my model declaration:

```python
    representation_model = OpenAI(client=openai_client, model="gpt-3.5-turbo", delay_in_seconds=10, chat=True)

    topic_model = BERTopic(
        embedding_model=embedding_model,
        zeroshot_topic_list=zeroshot_topic_list,
        zeroshot_min_similarity=.3,
        vectorizer_model=vectorizer_model,
        min_topic_size=2,
        nr_topics="auto",
        representation_model=representation_model
    )
    
    topics, _ = topic_model.fit_transform(documents=texts, embeddings=np.array(embeddings))
```
    
What am I doing wrong and how can I possibly overcome this error? Thanks!

**Comment by MaartenGr** [comment]
I'm missing a bit of information to get the complete picture. Could you share your **full** code and your **full** error message? That helps me understand how certain variables are created, the order of things, etc. What version of BERTopic are you using?

Lastly, how many documents are in `texts`? I assume this variable is the same as `secondary` even though they have different names, right?

**Reply by mzhadigerov** [reply]
No, `main` and `secondary` are two different lists, with different texts and different length. To be more specific: I parsed two pdf files and put texts from one pdf to `main` and texts from the second pdf to `secondary`.

```python
main = [...] # length == 53

# embeddings = I extract them from vector database

embedding_model = OpenAIBackend(...text embed 3 large...)
representation_model = OpenAI(...)
vectorizer_model = CountVectorizer(stop_words="english", ngram_range=(1, 3))


## Notice there is no zeroshot_topic_list. 
topic_model = BERTopic(
   embedding_model=embedding_model,
  zeroshot_min_similarity=.3,
  vectorizer_model=vectorizer_model,
  min_topic_size=2,
  nr_topics="auto",
  representation_model=representation_model
) 

topics, _ = topic_model.fit_transform(main, embeddings) 
topic_model.set_topic_labels(topic_model.topic_labels_) # <------ Up to this part everything works fine. I'm getting topics like ('Morgan Stanley Research on Securities and Investments: ', 'Financial Metrics Percentage Changes Over Time: ')

secondary = [...] # length == 87

# embeddings = I extract them from vector database

zeroshot_topic_list = topic_model.custom_labels_ # <--- Here I am appending labels from the previous model, to use them as zero-shot labels in the new model (e.g.: 'Morgan Stanley Research on Securities and Investments: ', 'Financial Metrics Percentage Changes Over Time: ')

## Notice there IS zeroshot_topic_list. I'm using topics generated by PREVIOUS model as a zeroshot list to THIS model
topic_model = BERTopic(
  embedding_model = embedding_model,
  zeroshot_topic_list=zeroshot_topic_list,
  zeroshot_min_similarity=.3,
  vectorizer_model=vectorizer_model,
  min_topic_size=2,
  nr_topics="auto",
  representation_model=representation_model
)

topics, _ = topic_model.fit_transform(secondary, embeddings)
```


I'm getting the error:

```
ERROR:    Exception in ASGI application
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/test/.pyenv/versions/3.10.12/lib/python3.10/concurrent/futures/process.py", line 246, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/home/test/AI/proj/myproj/app.py", line 277, in topic_modelling
    topics, _ = topic_model.fit_transform(documents=texts, embeddings=np.array(embeddings))
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/bertopic/_bertopic.py", line 408, in fit_transform
    umap_embeddings = self._reduce_dimensiontestty(embeddings, y)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/bertopic/_bertopic.py", line 3477, in _reduce_dimensiontestty
    umap_embeddings = self.umap_model.transform(embeddings)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/umap/umap_.py", line 2928, in transform
    raise ValueError(
ValueError: Transform unavailable when model was fit with only a single data sample.
"""
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 399, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
    return await self.app(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
  File "/home/test/AI/proj/myproj/venv/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/test/AI/proj/myproj/app.py", line 343, in run_topic_modelling
    raise e
  File "/home/test/AI/proj/myproj/app.py", line 337, in run_topic_modelling
    result = future.result()
  File "/home/test/.pyenv/versions/3.10.12/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/home/test/.pyenv/versions/3.10.12/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
ValueError: Transform unavailable when model was fit with only a single data sample.
```

I'm using `BERTopic v0.16.2`

**Reply by mzhadigerov** [reply]
@MaartenGr If I remove the `zeroshot_topic_list` (make it None or empty), everything works fine. 

With 'zeroshot_topic_list' the error is thrown by UMAP:

```python
        # If we fit just a single instance then error
        if self._raw_data.shape[0] == 1:
            raise ValueError(
                "Transform unavailable when model was fit with only a single data sample."
            )
  ```

**Reply by MaartenGr** [reply]
Hmmm, that's odd. It almost seems like that by setting `zeroshot_min_similarity=.3` all documents expect 1 are assigned to the zero-shot topics and that the one that is left is to be clustered. Could you try lowering `zero_shot_similarity`? I think that might solve your issue, especially if you set it at 0. 

This might also be caused by empty/short or perhaps even "strange" documents in `secondary` that differ greatly from all others.

**Reply by mzhadigerov** [reply]
Yes, it kinda "solved" the problem, but I guess setting `zeroshot_min_similarity=0` is not what we should do? It will "force" all the documents to be appended to the first string in the `zeroshot_topic_list`, right?

Can I conclude that If I'm getting the  `"Transform unavailable when model was fit with only a single data sample."` error, then, most probably the documents in the second list differ greatly from the documents from the first list and output something like: `No matching topics found`?

**Reply by mzhadigerov** [reply]
I'm wondering about this:

even though there are no documents that can be assigned to one of the zeroshot topics in the secondary list, why it doesn't detect other topics? 

I thought it will detect other topics If no document can be mapped to zeroshot list or am I wrong?

**Reply by MaartenGr** [reply, marked as answer]
> Can I conclude that If I'm getting the "Transform unavailable when model was fit with only a single data sample." error, then, most probably the documents in the second list differ greatly from the documents from the first list and output something like: No matching topics found?

It is quite the opposite. Almost all documents in `secondary` are matched with the topics you created from `main`. What happens is that there was just a single document not matched which was then put through the default BERTopic pipeline. See the entire process [here](https://maartengr.github.io/BERTopic/getting_started/zeroshot/zeroshot.html). In practice, you could also increase the `zeroshot_min_similarity` value to make sure that there isn't one document left but potentially multiple. 

> even though there are no documents that can be assigned to one of the zeroshot topics in the secondary list, why it doesn't detect other topics?

That's the thing, it is the opposite. Almost all documents are matched to the zeroshot topics expect one. That one will then be put through the default pipeline of SBERT -> UMAP -> HDBSCAN -> cTFIDF but because there is only one document, reducing dimensionality and clustering doesn't work. 

> I thought it will detect other topics If no document can be mapped to zeroshot list or am I wrong?

It will do that as long as there are at least 3 documents (the same value as `min_topic_size`) which in your case is unfortunately only 1.

**Reply by mzhadigerov** [reply]
@MaartenGr Aaaah! I understand now. Thank you!

fix from my side:
So in this case the only thing I can do is increase `zeroshot_min_similarity`, right?

fix from BERTopic library side:
Maybe it makes sense to move that single document to "Outliers (-1)" to overcome the issue or even discard it from BERTopic pipeline?

**Reply by MaartenGr** [reply]
> So in this case the only thing I can do is increase zeroshot_min_similarity, right?

Yes, I believe that would help. 

> Maybe it makes sense to move that single document to "Outliers (-1)" to overcome the issue or even discard it from BERTopic pipeline?

Not sure in all honesty since this happens very uncommonly that only a single document is passed to the HDBSCAN model. I believe it might even be related to your data (perhaps empty documents or documents that end up empty due to their size, etc.) that might impact this. There is a new implementation of zeroshot topic modeling coming that might alleviate this issue.

**Reply by mzhadigerov** [reply]
@MaartenGr Hi! Is there any updates regarding the new implementation of zeroshot topic modeling you mentioned? I'm getting the error again on another dataset and always changing the value of `zeroshot_min_similarity` is not an option, unfortunately.

**Reply by MaartenGr** [reply]
@mzhadigerov The new implementation has already merged into the main branch. Have you tried installing BERTopic from the main branch?

**Reply by mzhadigerov** [reply]
@MaartenGr I am using v0.16,2

**Reply by MaartenGr** [reply]
When you run `pip install bertopic` you are installing the latest release. The main branch holds new features/fixes that are not yet in the latest release, so that you are using v0.16.2 doesn't tell me whether you installed from the main branch. Please install BERTopic from the main branch and see if that fixes the issue.

**Reply by mzhadigerov** [reply]
@MaartenGr Ah. Yes, you are right. I'm installing by `pip install bertopic`. Not sure If it clones from the main branch or from another one. Let me make sure that I'm cloning from the main branch.

**Reply by MaartenGr** [reply]
@mzhadigerov `pip install` never clones from the main branch as it is a separate (typically manual) release. If you install from the main branch, you should have updates that cannot be found with `pip install bertopic`.
